packages/in_app_purchase/example/ios/in_app_purchase_pluginTests/ProductRequestHandlerTest.m

```diff
@@ -0,0 +1,180 @@
+    [self setValue:discount forKey:@"introductoryPrice"];
+    [self setValue:@"com.group" forKey:@"subscriptionGroupIdentifier"];
+  }
+  return self;
+}
+
+@end
+
+#pragma tests start here
+
+@interface ProductRequestHandlerTest : XCTestCase
+
+@end
+
+@implementation ProductRequestHandlerTest
+
+- (void)testSKProductSubscriptionPeriodStubToMap {
+  SKProductSubscriptionPeriodStub *period = [[SKProductSubscriptionPeriodStub alloc] init];
+  NSDictionary *map = [period toMap];
+  NSDictionary *match = @{@"numberOfUnits" : @(0), @"unit" : @(0)};
+  XCTAssertEqualObjects(map, match);
+
+  NSDictionary *notMatch = @{@"numberOfUnits" : @(0), @"unit" : @(1)};
+  XCTAssertNotEqualObjects(map, notMatch);
+}
+
+- (void)testSKProductDiscountStubToMap {
+  SKProductDiscountStub *discount = [[SKProductDiscountStub alloc] init];
+  NSDictionary *map = [discount toMap];
+  NSDictionary *match = @{
+    @"price" : @(1.0),
+    @"currencyCode" : @"USD",
+    @"numberOfPeriods" : @(1),
+    @"subscriptionPeriod" : @{@"numberOfUnits" : @(0), @"unit" : @(0)},
+    @"paymentMode" : @(1)
+  };
+  XCTAssertEqualObjects(map, match);
+
+  NSDictionary *notMatch = @{
+    @"price" : @(1.0),
+    @"currencyCode" : @"USD",
+    @"numberOfPeriods" : @(1),
+    @"subscriptionPeriod" : @{@"numberOfUnits" : @(0), @"unit" : @(0)},
+    @"paymentMode" : @(0)
+  };
+  XCTAssertNotEqualObjects(map, notMatch);
+}
+
+- (void)testProductToMap {
+  SKProductStub *product = [[SKProductStub alloc] init];
+  NSDictionary *map = [product toMap];
+  NSDictionary *match = @{
+    @"price" : @(1.0),
+    @"currencyCode" : @"USD",
+    @"productIdentifier" : @"consumable",
+    @"localizedTitle" : @"title",
+    @"localizedDescription" : @"description",
+    @"downloadable" : @YES,
+    @"downloadContentLengths" : @[ @1, @2 ],
+    @"downloadContentVersion" : [NSNull null],
+    @"subscriptionPeriod" : @{@"numberOfUnits" : @(0), @"unit" : @(0)},
+    @"introductoryPrice" : @{
+      @"price" : @(1.0),
+      @"currencyCode" : @"USD",
+      @"numberOfPeriods" : @(1),
+      @"subscriptionPeriod" : @{@"numberOfUnits" : @(0), @"unit" : @(0)},
+      @"paymentMode" : @(1)
+    },
+    @"subscriptionGroupIdentifier" : @"com.group"
+  };
+  XCTAssertEqualObjects(map, match);
+
+  NSDictionary *notMatch = @{
+    @"price" : @(1.0),
+    @"currencyCode" : @"USD",
+    @"productIdentifier" : @"consumable",
+    @"localizedTitle" : @"title",
+    @"downloadable" : @YES,
+    @"downloadContentLengths" : @[ @1, @2 ],
+    @"subscriptionPeriod" : @{@"numberOfUnits" : @(0), @"unit" : @(0)},
+    @"introductoryPrice" : @{
+      @"price" : @(1.0),
+      @"currencyCode" : @"USD",
+      @"numberOfPeriods" : @(1),
+      @"subscriptionPeriod" : @{@"numberOfUnits" : @(0), @"unit" : @(0)},
+      @"paymentMode" : @(1)
+    },
+    @"subscriptionGroupIdentifier" : @"com.group"
+  };
+  XCTAssertNotEqualObjects(map, notMatch);
+}
+
+- (void)testRequestDelegateSetup {
```

Comment:
Checking for if `delegate` is set seems like it's testing an implementation detail to me, do we need it? For this test to work it looks like a lot of private methods are being exposed just to check internal state on the class, which generally leads to brittle tests. It seems like the thing that really matters is it executes a request and passes a response when `startWithProductIdentifiers` is called. Can we test for that directly instead?

https://testing.googleblog.com/2013/08/testing-on-toilet-test-behavior-not.html